db/warehouse/migrate/20250121204933_add_hmis_household_events_table.rb

```diff
@@ -0,0 +1,29 @@
+#  Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+#  License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+#
+
+class AddHmisHouseholdEventsTable < ActiveRecord::Migration[7.0]
+  def change
+    create_table :hmis_household_events do |t|
+      t.string :household_id, index: true # hmis_households is a view, not a table
+      t.references :data_source, null: false
+      t.string :event_type, null: false
+      t.jsonb :event_details
+      t.references :user, null: false
+      t.timestamps
+    end
+
+    # Create a join table since Postgres can't put fk constraints on array elements
+    create_table :hmis_household_event_enrollments, id: false do |t|
+      t.references :hmis_household_event, null: false, foreign_key: { to_table: :hmis_household_events }, index: { name: 'idx_hmis_household_event_enrollments_on_event_id' }
+      t.references :enrollment, null: false, foreign_key: { to_table: :Enrollment }, index: true # this is the ID of the enrollment record: Enrollment.id not Enrollment.EnrollmentID
+    end
+
+    # Add a composite unique index to prevent duplicate associations
+    add_index :hmis_household_event_enrollments,
+              [:hmis_household_event_id, :enrollment_id],
+              unique: true,
+              name: 'idx_hmis_household_event_enrollments_on_event_and_enrollment'
+  end
```

Comment:
I'm wondering if this complexity is worth it just for fk constraints on the enrollment references.

For other similar tales (client merge histories, client merge audits) we don't have fk constraints.
The primary goal is to record events as they happen, even if the related records (Enrollments) are deleted or changed. Because of that, I think the fk constraints might not be necessary because the historical accuracy of the audit trail is more important than referential integrity.


For example if an Enrollment get's hard-deleted, we probably want the hmis_household_events that it was a part of to remain. We also don't want the soft-deletion of an enrollment to cascade and delete the household events that it was a part of.